File: server/app/domain/canonico.py
"""Serialización canónica y hash de payloads.

Esta es la pieza más delicada de tener dos lenguajes: el `hash_payload` que
calcula Dart y el que verifica Python deben coincidir **byte a byte**. Si no,
la alarma contra manipulación (mismo `operacion_id` con payload distinto) se
convierte en un generador de falsos positivos y se acaba desactivando.

Los cuatro lugares donde Dart y Python divergen en silencio están cerrados
por reglas explícitas, no por convención:

1. **Orden de claves** — se ordenan por punto de código Unicode.
2. **`Decimal` vs `double`** — los flotantes están PROHIBIDOS en el payload.
   El dinero y las cantidades viajan como *string* con escala fija.
3. **Fechas** — RFC 3339 en UTC, con exactamente 3 decimales y sufijo `Z`.
4. **`null` vs clave ausente** — las claves con valor nulo se OMITEN.

Módulo de dominio puro: sin imports de FastAPI ni de SQLAlchemy.
"""

from __future__ import annotations

import hashlib
import re
from collections.abc import Mapping, Sequence
from datetime import UTC, datetime
from decimal import Decimal
from typing import Any

__all__ = [
    "ESCALA_CANTIDAD",
    "ESCALA_DINERO",
    "PayloadNoCanonico",
    "a_texto_canonico",
    "formatear_cantidad",
    "formatear_dinero",
    "formatear_instante",
    "hash_payload",
]

ESCALA_DINERO = 2
ESCALA_CANTIDAD = 3

# 'dinero' y 'cantidad' se validan con estas formas exactas: sin separador de
# miles, sin '+', sin notación exponencial, con la escala fija completa.
_RE_DINERO = re.compile(rf"^-?\d+\.\d{{{ESCALA_DINERO}}}$")
_RE_CANTIDAD = re.compile(rf"^-?\d+\.\d{{{ESCALA_CANTIDAD}}}$")
_RE_INSTANTE = re.compile(r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z$")

_ESCAPES = {
    '"': '\\"',
    "\\": "\\\\",
    "\b": "\\b",
    "\f": "\\f",
    "\n": "\\n",
    "\r": "\\r",
    "\t": "\\t",
}


class PayloadNoCanonico(ValueError):
    """El payload contiene algo que no admite una representación canónica."""


def formatear_dinero(valor: Decimal | int | str) -> str:
    """Dinero como string con exactamente 2 decimales.

    Nunca aceptar `float` aquí: es justo el tipo que pierde centavos.
    """
    if isinstance(valor, float):
        raise PayloadNoCanonico("el dinero nunca se representa como float")
    d = Decimal(valor)
    return f"{d.quantize(Decimal('1.' + '0' * ESCALA_DINERO)):f}"


def formatear_cantidad(valor: Decimal | int | str) -> str:
    """Cantidad como string con exactamente 3 decimales (fracción de caja, granel)."""
    if isinstance(valor, float):
        raise PayloadNoCanonico("las cantidades nunca se representan como float")
    d = Decimal(valor)
    return f"{d.quantize(Decimal('1.' + '0' * ESCALA_CANTIDAD)):f}"


def formatear_instante(momento: datetime) -> str:
    """RFC 3339 en UTC con exactamente 3 decimales y sufijo 'Z'.

    Un datetime sin zona horaria es un error: el reloj del dispositivo y el del
    servidor están en juego, y una fecha ambigua entre ambos es indefendible.
    """
    if momento.tzinfo is None:
        raise PayloadNoCanonico("el instante debe traer zona horaria explícita")
    utc = momento.astimezone(UTC)
    return f"{utc.strftime('%Y-%m-%dT%H:%M:%S')}.{utc.microsecond // 1000:03d}Z"


def _escapar(texto: str) -> str:
    salida = ['"']
    for ch in texto:
        if ch in _ESCAPES:
            salida.append(_ESCAPES[ch])
        elif ch < "\x20":
            salida.append(f"\\u{ord(ch):04x}")
        else:
            # El resto se emite literal en UTF-8: nada de \uXXXX para acentos
            # ni emoji. Dart hace lo mismo por defecto.
            salida.append(ch)
    salida.append('"')
    return "".join(salida)


def _canonizar(valor: Any, ruta: str) -> str:
    if valor is True:
        return "true"
    if valor is False:
        return "false"
    if isinstance(valor, str):
        return _escapar(valor)
    if isinstance(valor, int):
        # bool ya quedó atrapado arriba. Los enteros van sin comillas.
        return str(valor)
    if isinstance(valor, float):
        raise PayloadNoCanonico(
            f"{ruta}: los float están prohibidos en el payload; "
            "usa string con escala fija (ver formatear_dinero/formatear_cantidad)"
        )
    if isinstance(valor, Decimal):
        raise PayloadNoCanonico(
            f"{ruta}: convierte el Decimal a string con escala fija antes de canonizar"
        )
    if isinstance(valor, datetime):
        raise PayloadNoCanonico(f"{ruta}: convierte el datetime con formatear_instante()")
    if isinstance(valor, Mapping):
        partes = []
        for clave in sorted(valor.keys()):
            if not isinstance(clave, str):
                raise PayloadNoCanonico(f"{ruta}: las claves deben ser string")
            interno = valor[clave]
            if interno is None:
                continue  # regla 4: las claves nulas se omiten
            partes.append(f"{_escapar(clave)}:{_canonizar(interno, f'{ruta}.{clave}')}")
        return "{" + ",".join(partes) + "}"
    if isinstance(valor, Sequence):
        # El orden de un arreglo SÍ es significativo: las partidas de una venta
        # llevan número de línea y el ticket impreso las muestra en ese orden.
        return "[" + ",".join(
            _canonizar(v, f"{ruta}[{i}]") for i, v in enumerate(valor)
        ) + "]"
    if valor is None:
        raise PayloadNoCanonico(f"{ruta}: null solo es válido como valor de una clave")
    raise PayloadNoCanonico(f"{ruta}: tipo no canonizable ({type(valor).__name__})")


def a_texto_canonico(payload: Mapping[str, Any]) -> str:
    """Forma canónica del payload. Determinista en Python y en Dart."""
    if not isinstance(payload, Mapping):
        raise PayloadNoCanonico("el payload raíz debe ser un objeto")
    return _canonizar(payload, "$")


def hash_payload(payload: Mapping[str, Any]) -> str:
    """SHA-256 hexadecimal en minúsculas de la forma canónica."""
    return hashlib.sha256(a_texto_canonico(payload).encode("utf-8")).hexdigest()


def validar_dinero(texto: str) -> str:
    if not _RE_DINERO.match(texto):
        raise PayloadNoCanonico(f"dinero mal formado: {texto!r} (se esperaba p. ej. '250.00')")
    return texto


def validar_cantidad(texto: str) -> str:
    if not _RE_CANTIDAD.match(texto):
        raise PayloadNoCanonico(f"cantidad mal formada: {texto!r} (se esperaba p. ej. '12.000')")
    return texto


def validar_instante(texto: str) -> str:
    if not _RE_INSTANTE.match(texto):
        raise PayloadNoCanonico(
            f"instante mal formado: {texto!r} (se esperaba '2026-09-15T03:14:07.123Z')"
        )
    return texto

File: mobile/packages/dsd_core/lib/src/canonico.dart
/// Serialización canónica y hash de payloads — mitad Dart del contrato.
///
/// Espejo de `server/app/domain/canonico.py`. Las reglas están en
/// `contracts/README.md` §1 y los vectores en `contracts/canonical_vectors.json`,
/// que esta implementación debe pasar íntegros.
///
/// **Trampa principal:** `String.compareTo` de Dart ordena por unidades UTF-16,
/// no por punto de código. Para claves fuera del plano básico (emoji) eso da un
/// orden DISTINTO al de Python: 'ﬀ' (U+FB00) va antes que '🔒' (U+1F512) por
/// punto de código, pero después por UTF-16. Por eso aquí se compara por runas.
/// El vector `orden_claves_fuera_del_plano_basico` fija justo ese caso.
///
/// Lo valida `test/canonico_test.dart` contra los vectores compartidos, que son
/// los MISMOS que ejecuta la suite de Python.
library;

import 'dart:convert';

import 'package:crypto/crypto.dart';

const int escalaDinero = 2;
const int escalaCantidad = 3;

class PayloadNoCanonico implements Exception {
  final String mensaje;
  PayloadNoCanonico(this.mensaje);
  @override
  String toString() => 'PayloadNoCanonico: $mensaje';
}

/// Compara por punto de código Unicode, como `sorted()` de Python.
int compararPorPuntoDeCodigo(String a, String b) {
  final ra = a.runes.toList();
  final rb = b.runes.toList();
  final n = ra.length < rb.length ? ra.length : rb.length;
  for (var i = 0; i < n; i++) {
    if (ra[i] != rb[i]) return ra[i] < rb[i] ? -1 : 1;
  }
  return ra.length.compareTo(rb.length);
}

/// Dinero como string con exactamente 2 decimales: '250.00', '-125.50'.
///
/// Recibe String o int a propósito: aceptar `double` aquí sería reintroducir
/// justo el tipo que pierde centavos. Usa el paquete `decimal` para operar.
String formatearDinero(Object valor) => _formatearEscalaFija(valor, escalaDinero, 'dinero');

/// Cantidad como string con exactamente 3 decimales: '12.000', '1.375'.
String formatearCantidad(Object valor) =>
    _formatearEscalaFija(valor, escalaCantidad, 'cantidad');

String _formatearEscalaFija(Object valor, int escala, String etiqueta) {
  if (valor is double) {
    throw PayloadNoCanonico('el $etiqueta nunca se representa como double');
  }
  final texto = valor.toString();
  final negativo = texto.startsWith('-');
  final absoluto = negativo ? texto.substring(1) : texto;
  final partes = absoluto.split('.');
  if (partes.length > 2 || partes[0].isEmpty || !RegExp(r'^\d+$').hasMatch(partes[0])) {
    throw PayloadNoCanonico('$etiqueta mal formado: $texto');
  }
  var decimales = partes.length == 2 ? partes[1] : '';
  if (!RegExp(r'^\d*$').hasMatch(decimales)) {
    throw PayloadNoCanonico('$etiqueta mal formado: $texto');
  }
  if (decimales.length > escala) {
    throw PayloadNoCanonico(
        '$etiqueta con más de $escala decimales: $texto (redondea antes)');
  }
  decimales = decimales.padRight(escala, '0');
  return '${negativo ? '-' : ''}${partes[0]}.$decimales';
}

/// RFC 3339 en UTC con exactamente 3 decimales y sufijo 'Z'.
String formatearInstante(DateTime momento) {
  final utc = momento.toUtc();
  String dos(int n) => n.toString().padLeft(2, '0');
  final ms = utc.millisecond.toString().padLeft(3, '0');
  return '${utc.year.toString().padLeft(4, '0')}-${dos(utc.month)}-${dos(utc.day)}'
      'T${dos(utc.hour)}:${dos(utc.minute)}:${dos(utc.second)}.${ms}Z';
}

String _escapar(String texto) {
  final b = StringBuffer('"');
  for (final unidad in texto.codeUnits) {
    switch (unidad) {
      case 0x22:
        b.write(r'\"');
      case 0x5C:
        b.write(r'\\');
      case 0x08:
        b.write(r'\b');
      case 0x0C:
        b.write(r'\f');
      case 0x0A:
        b.write(r'\n');
      case 0x0D:
        b.write(r'\r');
      case 0x09:
        b.write(r'\t');
      default:
        if (unidad < 0x20) {
          b.write('\\u${unidad.toRadixString(16).padLeft(4, '0')}');
        } else {
          // Las unidades suplentes se reemiten tal cual y el par se reconstruye
          // al codificar en UTF-8. Acentos y emoji van literales.
          b.writeCharCode(unidad);
        }
    }
  }
  b.write('"');
  return b.toString();
}

String _canonizar(Object? valor, String ruta) {
  if (valor is bool) return valor ? 'true' : 'false';
  if (valor is String) return _escapar(valor);
  if (valor is int) return valor.toString();
  if (valor is double) {
    throw PayloadNoCanonico(
        '$ruta: los double están prohibidos en el payload; usa string con escala fija');
  }
  if (valor is DateTime) {
    throw PayloadNoCanonico('$ruta: convierte el DateTime con formatearInstante()');
  }
  if (valor is Map) {
    final claves = valor.keys.map((k) {
      if (k is! String) throw PayloadNoCanonico('$ruta: las claves deben ser String');
      return k;
    }).toList()
      ..sort(compararPorPuntoDeCodigo);
    final partes = <String>[];
    for (final clave in claves) {
      final interno = valor[clave];
      if (interno == null) continue; // regla 2: las claves nulas se omiten
      partes.add('${_escapar(clave)}:${_canonizar(interno, '$ruta.$clave')}');
    }
    return '{${partes.join(',')}}';
  }
  if (valor is List) {
    // El orden de un arreglo SÍ es significativo.
    final partes = <String>[];
    for (var i = 0; i < valor.length; i++) {
      partes.add(_canonizar(valor[i], '$ruta[$i]'));
    }
    return '[${partes.join(',')}]';
  }
  if (valor == null) {
    throw PayloadNoCanonico('$ruta: null solo es válido como valor de una clave');
  }
  throw PayloadNoCanonico('$ruta: tipo no canonizable (${valor.runtimeType})');
}

/// Forma canónica del payload. Determinista en Dart y en Python.
String aTextoCanonico(Map<String, Object?> payload) => _canonizar(payload, r'$');

/// SHA-256 hexadecimal en minúsculas de la forma canónica.
String hashPayload(Map<String, Object?> payload) =>
    sha256.convert(utf8.encode(aTextoCanonico(payload))).toString();
